### ISSUE TITLE
Rule to require defining prototype functions with Object Notation

### ISSUE THREAD

**Issue by TheSavior** [eslint/eslint#3285]
I would like to have a rule where the following are warnings:

```
function MyObj() {
}

MyObj.prototype.foo = function() {
  
};

MyObj.prototype.bar = function() {
  
};

String.prototype.trim = function() {
};
```

And the following is considered valid:
```
function MyObj() {
}

MyObj.prototype = {
  foo: function() {

  },

  bar: function() {
  
  }
}

Object.assign(String.prototype, {
  trim: function() {}
});
```

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by michaelficarra** [comment]
What's the motivation behind this?

**Comment by TheSavior** [comment]
Readability, preference, consistency. Same as most rules. :)

**Comment by mysticatea** [comment]
It looks anti-patterns to me.

Because:

- `MyObj.prototype.constructor` becomes wrong object.
- To add enumerable properties into the builtin prototypes confuse `for-in` loops.

**Comment by TheSavior** [comment]
@mysticatea Can you elaborate on your points. What do you mean by "wrong object" and "confuses `for-in` loops"?

**Comment by michaelficarra** [comment]
@TheSavior You're setting the prototype object, losing the original `prototype.constructor` reference. And you're adding enumerable fields to the prototype, so for-in will iterate over them, something most people would not expect. These problems are the reason why you're one of the few people that prefer this pattern.

**Comment by ilyavolodin** [comment]
I always used this style with Backbone, due to the fact that Backbone uses _.extends for inheritance and doing a `for-in` on the model is a bad idea to begin with. So I though that I could use a rule like that. But now I actually think it should not be in the core. It's a good use-case for a plugin though.

**Comment by nzakas** [comment]
Agree, I think it's better to write a custom rule.
